python_modules/dagster/dagster/_core/pipes/utils.py

```diff
@@ -299,12 +333,65 @@ def _reader_thread(
             time.sleep(1)
 
 
+class PipesBlobStoreStdioReader(ABC):
+    def __init__(self, *, interval: float = 10, target_stream: TextIO):
+        self.interval = interval
+        self.target_stream = target_stream
+        self.thread: Optional[Thread] = None
+
+    @abstractmethod
+    def download_log_chunk(self, params: PipesParams) -> Optional[str]: ...
+
+    @abstractmethod
+    def log_exists(self, params: PipesParams) -> bool: ...
+
+    def start(self, params: PipesParams, is_task_complete: Event) -> None:
+        self.thread = Thread(target=self._reader_thread, args=(params, is_task_complete))
+        self.thread.start()
+
+    def stop(self) -> None:
+        if self.thread:
+            self.thread.join()
+
+    def _reader_thread(
+        self,
+        params: PipesParams,
+        is_task_complete: Event,
+    ) -> None:
+        start_or_last_download = datetime.datetime.now()
+        while True:
+            now = datetime.datetime.now()
+            if (now - start_or_last_download).seconds > self.interval or is_task_complete.is_set():
+                start_or_last_download = now
+                chunk = self.download_log_chunk(params)
+                if chunk:
+                    self.target_stream.write(chunk)
+                elif is_task_complete.is_set():
+                    break
+            time.sleep(self.interval)
+
+
+class PipesNoOpStdioReader(PipesBlobStoreStdioReader):
```

Comment:
This should inherit from a common pure interface, not the `PipesBlobStoreStdioReader` that has a bunch stuff in it that will never get called.

Reply:
I really feel that this is starting to overcomplicate the class hierarchy. We've dropped a few null checks, where IMO a `None` value is more intuitively obvious for functionality (i.e. this doesn't read stdout/stderr at all), for a special NoOp class and now a one-off interface _just_ to abstract out the common part between the NoOp and the (already abstract) variant that actually works?

I feel like I might be missing a broader vision here, do you see this interface being used elsewhere? And what should the name be? I would guess `PipesBlobStoreStdioReader` would be the name of the interface, but then the working variant becomes... `PipesBlobStoreThreadedStdioReader`?

Also, looking closely, I don't think in the current implementation there is anything in `PipesBlobStoreStdioReader` that never gets called for no-op. Instead the no-op just always returns `None` for the log chunk.

Reply:
Another point, I think probably we're going to want to restructure so there is just a top-level "Reader" class that has a list of component readers (messages, stdout, stderr, maybe even other logs), rather than the current scenario where message reading is in the top-level but stderr/stdout is a component. So maybe hold off on more refactoring here for the moment?